src/connection.js

```diff
@@ -0,0 +1,112 @@
+const { Map, fromJS } = require('immutable')
+const OpSet = require('./op_set')
+
+// Returns true if all components of clock1 are less than or equal to those of clock2.
+// Returns false if the clocks are incomparable or if clock1 is greater than clock2.
+function lessOrEqual(clock1, clock2) {
+  return clock1.keySeq().concat(clock2.keySeq()).reduce(
+    (result, key) => (result && clock1.get(key, 0) <= clock2.get(key, 0)),
+    true)
+}
+
+function mergeClock(clockMap, docId, clock) {
+  clock = clockMap.get(docId, Map()).mergeWith((x, y) => Math.max(x, y), clock)
+  return clockMap.set(docId, clock)
+}
+
+// Keeps track of the communication with one particular peer. Allows updates for many documents to
+// be multiplexed over a single connection.
+//
+// To integrate a connection with a particular networking stack, two functions are used:
+// * `sendMsg` (callback passed to the constructor, will be called when local state is updated)
+//   takes a message as argument, and sends it out to the remote peer.
+// * `receiveMsg` (method on the connection object) should be called by the network stack when a
+//   message is received from the remote peer.
+//
+// The documents to be synced are managed by a `DocSet`. Whenever a document is changed locally,
```

Comment:
Let's talk about docId soon. It's a gnarly hack that I'm concerned will become too ensconced to excise conveniently.

Reply:
Yes, I was a little concerned about docId too — let me know what medium you'd like to use to talk.

Reply:
I haven't got a specific proposal yet but I hate to see it infecting automerge. Can we push it up into MPL? Can we replace docId with a convention where all documents share a namespace but begin under a docId field? (IOW you might find the title for a document at store[docId][title]).

Reply:
Hmm, not sure about that convention — I think it would interact weirdly with the fact that an Automerge document is immutable. The role of DocSet here is to broker a compromise between the immutability of Automerge core, and the fact that we need to process mutations that come in from the network. I thought the cleanest way of handling state changes due to network activity was to make DocSet and Connection mutable, while keeping the Automerge core immutable.

If we put all the documents in a shared namespace, that would probably mean that the top-level namespace must be mutable, while the documents within it would remain immutable. And I think that would be more confusing than helpful.

Alternatively, it might be possible to restructure the API so that every message from the network goes through the same event loop as local user input. That way, the Automerge.Connection abstraction could also be made immutable. But I'm not convinced that would be better… a network connection is essentially a kind of state machine, with state transitions linked to sending and receiving of messages, and I think that kind of pattern is more naturally handled in a mutable way.